File: problems/06_linked_list/0138_copy_list_with_random_pointer/solution.py
"""
Problem: 0138. Copy List with Random Pointer
Difficulty: Medium
URL: https://leetcode.com/problems/copy-list-with-random-pointer/

解法：HashMap 兩遍掃描，O(n)/O(n)。
"""

from __future__ import annotations

from typing import Optional

class Node:
    def __init__(self, x: int, next: Optional["Node"] = None, random: Optional["Node"] = None):
        pass


def build(arr: list[tuple[int, int | None]]) -> Optional[Node]:
    if not arr:
        return None
    nodes = [Node(v) for v, _ in arr]
    for i in range(len(nodes) - 1):
        nodes[i].next = nodes[i + 1]
    for i, (_, r) in enumerate(arr):
        if r is not None:
            nodes[i].random = nodes[r]
    return nodes[0]

def serialize(h: Optional[Node]) -> list[tuple[int, int | None]]:
    nodes, idx = [], {}
    cur = h
    while cur:
        idx[id(cur)] = len(nodes)
        nodes.append(cur)
        cur = cur.next
    out = []
    for n in nodes:
        out.append((n.val, idx[id(n.random)] if n.random else None))
    return out

class Solution:
    def copyRandomList(self, head: Optional[Node]) -> Optional[Node]:
        pass


if __name__ == "__main__":
    node = Node()

    # Test 1
    # result = node.solve(...)
    # assert result == expected

    print("All tests passed!")

File: problems/06_linked_list/0138_copy_list_with_random_pointer/solution.dart
// Problem: 0138. Copy List with Random Pointer
// Difficulty: Medium

class RNode {
  int val;
  RNode? next;
  RNode? random;
  RNode(this.val);
}

RNode? build(List<List<int?>> arr) {
  if (arr.isEmpty) return null;
  final nodes = arr.map((p) => RNode(p[0]!)).toList();
  for (var i = 0; i < nodes.length - 1; i++) nodes[i].next = nodes[i + 1];
  for (var i = 0; i < arr.length; i++) {
    final r = arr[i][1];
    if (r != null) nodes[i].random = nodes[r];
  }
  return nodes[0];
}

List<List<int?>> serialize(RNode? h) {
  final nodes = <RNode>[];
  final idx = <RNode, int>{};
  var c = h;
  while (c != null) { idx[c] = nodes.length; nodes.add(c); c = c.next; }
  return nodes.map((n) => <int?>[n.val, n.random == null ? null : idx[n.random!]]).toList();
}

RNode? copyRandomList(RNode? head) {
  if (head == null) return null;
  final m = <RNode, RNode>{};
  var c = head;
  while (true) {
    m[c] = RNode(c.val);
    if (c.next == null) break;
    c = c.next!;
  }
  var cur = head;
  while (true) {
    final n = m[cur]!;
    n.next = cur.next == null ? null : m[cur.next!];
    n.random = cur.random == null ? null : m[cur.random!];
    if (cur.next == null) break;
    cur = cur.next!;
  }
  return m[head];
}

bool _eq(List<List<int?>> a, List<List<int?>> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i][0] != b[i][0] || a[i][1] != b[i][1]) return false;
  }
  return true;
}

void main() {
  final cases = <List<List<int?>>>[
    [[7, null], [13, 0], [11, 4], [10, 2], [1, 0]],
    [[1, 1], [2, 1]],
    [[3, null], [3, 0], [3, null]],
    [],
  ];
  for (final arr in cases) {
    final h = build(arr);
    final c = copyRandomList(h);
    if (!_eq(serialize(c), arr)) throw StateError('FAIL serialize');
  }
  print('All tests passed!');
}
